app: Add `rpm-ostree compose extensions`

jlebon opened the pull request:
This adds support for a new `rpm-ostree compose extensions` command`
which takes a treefile, a new extensions YAML file, and an OSTree repo
and ref. It performs a depsolve and downloads the extensions to a
provided output directory.

This is intended to replace cosa's `download-extensions`:
coreos/coreos-assembler@master/src/download-extensions

The input YAML schema matches the one accepted by that script.

Some differences from the script:
- We have a guaranteed depsolve match and thus can avoid silly issues
  we've hit in RHCOS (like downloading the wrong `libprotobuf` for
  `usbguard` -- rhbz#1889694).
- We seamlessly re-use the same repos defined in the treefile, whereas
  the cosa script uses `reposdir=$dir` which doesn't have the same
  semantics (repo enablement is in that case purely based on the
  `enabled` flag in those repos, which may be different than what the
  rpm-ostree compose ran with).
- We perform more sanity-checks against the requested extensions, such
  as whether the extension is already in the base.
- We support no-change detection via a state SHA512 file for better
  integration in cosa and pipelines.
- We support a `match-base-evr` key, which forces the extension to have
  the same EVR as the one from a base package: this is helpful in the
  case of extensions which complement a base package, esp. those which
  may not have strong enough reldeps to enforce matching EVRs by
  depsolve alone (`kernel-headers` is an example of this).
- We don't try to organize the RPMs into separate directories by
  extension because IMO it's not at the right level. Instead, we should
  work towards higher-level metadata to represent extensions (see
  openshift/os#409 which is related to this).

Closes: #2055

jlebon commented:
Example:

```diff
$ git diff
diff --git a/manifests/fedora-coreos.yaml b/manifests/fedora-coreos.yaml
index c92effc..0f435aa 100644
--- a/manifests/fedora-coreos.yaml
+++ b/manifests/fedora-coreos.yaml
@@ -23,6 +23,9 @@ packages:
   # Updates
   - zincati

+package-extensions:
+  - ltrace
+
 etc-group-members:
   # Add the docker group to /etc/group
   # https://github.com/coreos/fedora-coreos-tracker/issues/2
```

```
$ rpm-ostree compose extensions --repo tmp/repo --cachedir cache --base-rev fedora/x86_64/coreos/testing-devel src/config/manifest.yaml  --output-dir extensions
Checking out 0f1245b... done!
warning: Found bdb Packages database while attempting sqlite backend: using bdb backend.
Enabled rpm-md repositories: fedora fedora-updates
rpm-md repo 'fedora' (cached); generated: 2020-10-19T23:27:19Z
rpm-md repo 'fedora-updates' (cached); generated: 2021-01-11T00:54:51Z
⠁ Importing rpm-md   0% [░░░░░░░░░░░░░░░░░░░░] (0s)
Importing rpm-md... done
Resolving dependencies... done
Will download: 1 package (143.2 kB)
Downloading from 'fedora'... done
$ ll extensions
total 140
-rw-r--r--. 1 jlebon jlebon 143159 Jan 11 16:58 ltrace-0.7.91-39.fc33.x86_64.rpm
```

(As expected, this runs fine unprivileged.)

I tried to write this up in Rust via cxxbridge, though the actual amount of novel code here is minimal. It's all just re-using pieces we already have like `RpmOstreeContext` and the `Treefile` object.

One thing I investigated was to make `Treefile` more usable via the bridge, but that requires moving out the interesting `impl Treefile` methods so that they're bridgeable, and then you also lose a lot of nice typing on the parameters in the process of making it cxx.rs-compatible (like `Path` and `openat::Dir`). I think the better approach is doing https://github.com/coreos/rpm-ostree/pull/2413 so that we can call e.g. `core.cxx` code from Rust, so then we should be able to write up this kind of glue code in Rust instead.

jlebon commented:
I also went back and forth on whether to put this in the treefile, or have a separate `extensions.yaml` file. Since this is a separate command, it feels a bit odd to add a field there which `rpm-ostree compose tree` doesn't really care about (theoretically in the future, it could also take care of downloading extensions, though as mentioned in https://github.com/coreos/rpm-ostree/issues/2055, it's not that simple). It does feel more natural though for OS maintainers to keep it in the same file (and plus, we get `releasever` and `repos` which we also need for extensions).

Hmm, actually... I just realized this will break hash comparisons for the OSTree compose change detection. So we'd have to also e.g. exclude that field from the hash, which is yucky. I think I'm leaning now towards making it a separate file instead which is basically a subset of the treefile schema.

lucab commented on rust/src/treefile.rs:
```diff
@@ -115,6 +115,10 @@ fn treefile_parse_stream<R: io::Read>(
         }
     }
 
+    if let Some(pkgs) = treefile.package_extensions.take() {
```
Can this be turned into a `if let Some(ref pkgs) = ...`?

cgwalters commented:
Yeah...tricky.  Perhaps separate command and file, but inherit things like `repos` from `manifest.yaml` by default?

Not necessary for the initial cut here, but let's think about correctly handling change detection in the future; a brutally hacky way to do this would be to just do the operation in a tempdir and then compare the repodata or something?

jlebon commented:
OK, rebased and fleshed out now! This is ready for review now and working pretty well, though I haven't yet tried integrating it into cosa.

cgwalters commented on rust/src/extensions.rs:
```diff
@@ -0,0 +1,191 @@
+pub struct Extension {
+    packages: Vec<String>,
+    architectures: Option<Vec<String>>,
+    #[serde(rename = "match-base-evr")]
```
I generally prefer using `#[serde(rename-all = "kebab-case")]` on the whole struct so it's enforced consistent.

cgwalters commented on rust/src/extensions.rs:
```diff
@@ -0,0 +1,191 @@
+            .unwrap_or(true)
+    });
+
+    let base_pkgs: HashMap<String, String> =
```
I think this could be a `HashMap<&str, &str>` at least and point into the `base_pkgs` vec to avoid string duplication.

cgwalters commented on rust/src/extensions.rs:
```diff
@@ -0,0 +1,191 @@
+            .collect()
+    }
+
+    pub(crate) fn state_checksum_changed(&self, chksum: &str, output_dir: &str) -> Result<bool> {
```
Anything exposed via cxx currently needs to go via `CxxResult`, see https://github.com/coreos/rpm-ostree/pull/2451

Hmm maybe this would be less of a trap if we specified `CxxResult` in `lib.rs` too...

cgwalters replied:
(I tried that, doesn't work)

cgwalters commented on rust/src/extensions.rs:
```diff
@@ -0,0 +1,191 @@
+
+    pub(crate) fn state_checksum_changed(&self, chksum: &str, output_dir: &str) -> Result<bool> {
+        let output_dir = openat::Dir::open(output_dir)?;
+        if let Some(mut f) = output_dir.open_file_optional(RPMOSTREE_EXTENSIONS_STATE_FILE)? {
```
(We should probably add `read_optional_file_to_string()` to openat-ext)

jlebon replied:
https://github.com/cgwalters/openat-ext/pull/18

cgwalters commented on src/app/rpmostree-compose-builtin-tree.cxx:
```diff
@@ -1430,3 +1441,164 @@ rpmostree_compose_builtin_tree (int             argc,
+  if (!packages)
+      return FALSE;
+
+  auto packages_mapping = std::make_unique<rust::Vec<rpmostreecxx::StringMapping>>();
```
The string duping here is unfortunate (not that I care about performance for this much, but we should know how to do this better in other places).  Hmm...  Hard to solve without having a Rust API for `rpm_ostree_db_query_all()` which itself blocks on the Rust -> C++ calls.

jlebon replied:
Yeah... it hurt a bit to write that.

One thing I tried doing was to hook up the bindings for `GPtrArray` -> `PtrArray` like you did for `GCancellable`, but I couldn't get it to work.

cgwalters commented on src/app/rpmostree-compose-builtin-tree.cxx:
```diff
@@ -1430,3 +1441,164 @@ rpmostree_compose_builtin_tree (int             argc,
+
+  g_print ("Checking out %.7s... ", base_rev);
+  OstreeRepoCheckoutAtOptions opts = { .mode = OSTREE_REPO_CHECKOUT_MODE_USER };
+  if (!ostree_repo_checkout_at (repo, &opts, cachedir_dfd, TMP_EXTENSIONS_ROOTFS,
```
Sort of tangentially related discussion...do you see us running this in cosa via supermin or not?  I think that influences checkouts of the tree because if we don't do supermin then the cache ostreerepo may be on the `cache.qcow2`.  Perhaps the simplest is to supermin it, but it's unfortunate because we don't actually need privileges.

jlebon replied:
Yeah. Hmm... maybe we could decouple the cachedir a bit more so it's available over 9p and so not part of the qcow2 while the workdir would still be on a POSIX fs in the VM. That will make composes slightly slower though because we'd no longer be able to use devino caching. But it'd allow e.g. `cosa fetch` and this code to run fully unprivileged. Not sure, probably not worth the trouble short-term.

cgwalters reviewed (commented):
Nice work!  Looks great overall.

> - We don't try to organize the RPMs into separate directories by
  extension because IMO it's not at the right level. Instead, we should
  work towards higher-level metadata to represent extensions (see
  openshift/os#409 which is related to this).

But isn't the MCO relying on that today?

jlebon commented:
> But isn't the MCO relying on that today?

AFAICT, the MCO is just hardcoding the list of supported extensions today and points rpm-ostree to it via a yumrepo to the mounted oscontainer. Part of what I'm proposing in https://github.com/openshift/os/issues/409#issuecomment-764977556 is to put the extensions JSON metadata in there too and have the MCO parse that to get the mapping of "extension name" -> "list of packages to `rpm-ostree install`".

jlebon commented:
Updated!

cgwalters commented:
OK I think this is overall pretty good but:

- Tests: Maybe 
- Docs: I am terrible about this myself, and @travier I think has been a good influence in pushing us towards better docs.  Since this isn't listed as experimental we should document it.  (Obviously our whole buildsystem stuff could be better doc'd).  This can come after merge and we can all do it collaboratively too.  Maybe to start we dump the content of the PR text in `docs/extensions.md` or something? 

Or another model could be: Before a release we dump all the docs in a hackmd and collectively iterate on improving/tweaking them and land that as a co-authored PR (instead of the model where one person writes and another does nitpicks, much cheaper to do that stuff interactively)

jlebon commented:
OK right, testing reveals this doesn't handle local repos correctly.
/hold

jlebon commented:
OK, updated this now:
- reverted https://github.com/coreos/rpm-ostree/commit/6a3e3d807d671a3c437fb60dc4f04ccd90e4adc6 since it was misguided
- added docs
- added compose test

jlebon commented:
/hold cancel

cgwalters reviewed (approved):
/lgtm
